nifi-nar-bundles/nifi-jmx-bundle/nifi-jmx-processors/src/main/java/org/apache/nifi/processors/jmx/GetJMX.java

```diff
@@ -0,0 +1,385 @@
+        } catch( InstanceNotFoundException|IntrospectionException|ReflectionException e ) {
+            getLogger().error( "Exception with MBean Server: " + e.getMessage() );
+        } catch( Exception e ) {
+            getLogger().error( "Exception performing MBean Query: " + e.getMessage() );
+        }
+
+        return metricsSet;
+    }
+
+
+
+
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession session) throws ProcessException {
+
+        if (queue.size() < 100) {
+            final long pollingMillis = context.getProperty(POLLING_INTERVAL).asTimePeriod(TimeUnit.SECONDS) * 1000;
+
+            if ((queueLastUpdated.get() < System.currentTimeMillis() - pollingMillis) && listingLock.tryLock()) {
+                try {
+                    final Set<Record> metrics = performQuery( context );
+
+                    queueLock.lock();
+
+                    try {
+                        metrics.removeAll(inProcess);
+                        metrics.removeAll(recentlyProcessed);
+
+                        queue.clear();
+                        queue.addAll(metrics);
+
+                        queueLastUpdated.set(System.currentTimeMillis());
+                        recentlyProcessed.clear();
+
+                        if (metrics.isEmpty()) {
+                            context.yield();
+                        }
+                    } finally {
+                        queueLock.unlock();
+                    }
+                } finally {
+                    listingLock.unlock();
+                }
+            }
+        }
+
+        final int batchSize = context.getProperty(BATCH_SIZE).asInteger();
+        final List<Record> jmxList = new ArrayList<>(batchSize);
+        queueLock.lock();
+        try {
+            queue.drainTo(jmxList, batchSize);
+            if (jmxList.isEmpty()) {
+                return;
+            } else {
+                inProcess.addAll(jmxList);
+            }
+        } finally {
+            queueLock.unlock();
+        }
+
+        final ListIterator<Record> itr = jmxList.listIterator();
+
+        FlowFile flowFile = null;
+        try {
+            while (itr.hasNext()) {
+                final Record record = itr.next();
+
+                flowFile = session.create();
+                flowFile = session.write(flowFile,new OutputStreamCallback() {
+                    @Override
+                    public void process(final OutputStream out) throws IOException {
+                        try {
+                            out.write(record.toString().getBytes());
+                        } catch (Exception e) {
+                            getLogger().error("Exception writing metric record to flowfile: " + e.getMessage() +
+                                    ": record content: [" + record.toString() + "]" );
+                        }
+                    }
+                });
+
+                session.transfer(flowFile, REL_SUCCESS);
```

Comment:
I usually assume that it depends on the design chosen, how useful they are for downstream processors, the total number of attributes and more importantly, the size/lenght of the attributes.

If you are likely to use it for routing the flowfile, than I reckon they can be added directly, instead of requiring the use to read a large JSON (i.e. payload * BATCH_SIZE), which I suspect would happen in memory.

ListenSyslog in batching mode should be a good example of a batched + attributes processor.

Cheers